### ISSUE TITLE
RFC: workflow for library publishing

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#1146]
Here's a simple proposal @aryairani and I were just discussing for how library management could be handled nicely. Writeup is still a bit rough but we wanted to capture.

- Arya has a repo on GitHub with all libraries he maintains. At the top level of this repo he may have subnamespaces corresponding to multiple libraries, such as `mylib`
- Initially, Arya does a bunch of development in `mylib`, or maybe `mylib.master`.
- When he's ready to publish an initial release, he does:

  ```ucm
  .mylib> move.namespace .mylib .mylib.master
  .mylib> fork master v1
  .mylib> cd v1
  ```

  or

  ```ucm
  .mylib.master> fork .mylib.master .mylib.v1
  ```

  respectively, followed by `push git@github.com/aryairani/lib .mylib.v1` followed by tweets.

- To do work on a new release of a library, Arya just continues hacking on `mylib.master` after forking off `v1`.
  - Every time nothing `todo` (and maybe when work on unrelated item starts), UCM does: `delete.patch mylib.v2.patch`
  - Arya's ready to release `v2`, does `fork master v2`, then `publish-update mylib.v1 mylib.v2`, UCM does the following:
    - Creates a composite patch out of the history of `mylib.v2.patch` back through the namespace hash of `mylib.v1`, puts that in `mylib.v2.patches.v1-to-v2`.
    - Arya pushes to GitHub, announces `v2` is ready on Twitter, etc. The published `v2` has a `patches` directory with `v1-to-v2` in it which users of `v1` can use to upgrade.
  - Arya does `fork master v3` to produce the next release and so on.
  - Arya could `publish-update mylib.v1 mylib.master`, then `fork `mylib.master mylib.v2`. If this is followed as a convention, then the full patch series will accumulate under `patches` of the published library versions

So over time, Arya will have `mylib.v1`, `mylib.v2`, etc, and `mylib.master` for current work 
One nice property is that even if someone can come in after the fact and produce these patches and this conventional repository structure.

**Comment by mitchellwrosen** [comment]
I like this a lot. One advantage git has here, though, is that you can just reference a commit with a name; it would be a lot nicer if you could just tag a namespace `v1` and then get it out of your way until you're ready to dig it up to produce a patch. With the proposal above, the old library versions have to hang around somewhere in the current namespace.

**Comment by pchiusano** [comment]
@mitchellwrosen yeah, we could really use a story for "archiving" namespaces so they aren't loaded by default but are perhaps still available to commands like `fork`.

**Comment by aryairani** [comment]
Re. git: granted, git branches and tags exist in a completely different universe from package / module names.  I could imagine using a more git-like convention where you `.mylib> fork master .tags.mylib.v1`, which kind of gets it out of your way.  What ucm wouldn't currently do that git does is prevent you from changing the contents of `.tags.mylib.v1`; I guess that makes it more like a git branch than a git tag, hmm....
